File: 백준/Gold/1956. 운동/운동.py
# 1956번 운동

import sys
input = sys.stdin.readline

INF = 10**9

V, E = map(int, input().strip().split())
dist = [[INF] * (V+1) for _ in range(V+1)]

for _ in range(E):
    a, b, c = map(int, input().strip().split())
    dist[a][b] = c

for k in range(1, V+1):
    for i in range(1, V+1):
        for j in range(1, V+1):
            dist[i][j] = min(dist[i][j], dist[i][k]+dist[k][j])

val = INF

for i in range(1, V+1):
    for j in range(1, V+1):
        if i != j :
            if dist[i][j] < INF and dist[j][i] < INF:
                val = min(val, dist[i][j]+dist[j][i])

if val == INF:
    print(-1)
else:
    print(val)

File: 백준/Gold/1956. 운동/운동.cc
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;
int INF = 100000000;
int V, E;

int main() {
    cin >> V >> E;
    
    vector<vector<int>> dist;
    dist.resize(V + 1, vector<int>(V + 1, INF));

    for (int i = 0; i < E; i++) {
        int a, b, c;
        cin >> a >> b >> c;
        dist[a][b] = c;
    }

    for (int k = 1; k <= V; k++) {
        for (int i = 1; i <= V; i++) {
            for (int j = 1; j <= V; j++) {
                dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);
            }
        }
    }
    int ans = INF;

    for (int i = 1; i <= V; i++) {
        for (int j = 1; j <= V; j++) {
            if (i != j) {
                if (dist[i][j] < INF && dist[j][i] < INF) {
                    ans = min(ans, dist[i][j] + dist[j][i]);
                }
            }
        }
    }

    if (ans != INF) {
        cout << ans;
    }
    else {
        cout << -1;
    }
    return 0;
}
